### PR TITLE
feat: .nuxtignore

### PR DESCRIPTION
**Pull request by clarkdo**
## Types of changes

- [x] New feature (a non-breaking change which adds functionality)

## Description

First prototype of ignore features.
Related: https://github.com/nuxt/rfcs/issues/21

- deprecate current ignore and ignorePrefix (Maybe in 3.0 ?), then support node-glob-gitignore 
- [x] same rules as `.gitignore/.eslintignore`
- [x] support `.nuxtignore` file under `srcDir`
- [x] support ignore pages/layouts to be ignored
- [x] support store
- [x] support middleware
- [x] ~~support plugin~~ This is unnecessary since `plugin` is already bundled on demand
- [x] watch .nuxtignore in dev mode

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by ricardogobbosouza** [comment]
Hi @clarkdo 
Could not you use this to bypass the other files in directories `assets`, `components` , `middleware`, `plugins`, `static`, `store` ?

**Comment by clarkdo** [comment]
@ricardogobbosouza Thanks for the reviewing, the features you mentioned are already on the plan, we'll finish them step by step.

**Comment by ricardogobbosouza** [comment]
@clarkdo  What do you think of moving `packages/builder/src/ignore.js` to `@nuxtjs/utils` ?

File: packages/builder/package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "fs-extra": "^7.0.1",
     "glob": "^7.1.3",
     "hash-sum": "^1.0.2",
+    "ignore": "^5.0.4",
```
Comment (ricardogobbosouza) [inline packages/builder/package.json]: Must be in `@nuxt/utils`

**Review by ricardogobbosouza — CHANGES_REQUESTED** [review]

File: packages/builder/src/builder.js [hunk 2]
```diff
@@ -332,10 +348,8 @@ export default class Builder {
     } else if (this._nuxtPages) {
       // Use nuxt.js createRoutes bases on pages/
       const files = {}
-        ; (await glob(`${this.options.dir.pages}/**/*.{${this.supportedExtensions.join(',')}}`, {
-        cwd: this.options.srcDir,
-        ignore: this.options.ignore
-      })).forEach((f) => {
+      const pages = await this.resolveFiles(this.options.dir.pages)
+      pages.forEach((f) => {
```
Comment (pi0) [inline packages/builder/src/builder.js]: Can we use `for of` here?

File: packages/builder/src/builder.js [hunk 3]
```diff
@@ -377,9 +391,13 @@ export default class Builder {
     // -- Store --
     // Add store if needed
     if (this.options.store) {
+      templateVars.stores = await this.resolveRelative(this.options.dir.store)
```
Comment (pi0) [inline packages/builder/src/builder.js]: stores => storeModules?

File: packages/utils/src/ignore.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import path from 'path'
+import fs from 'fs-extra'
+import ignore from 'ignore'
+
+export class Ignore {
```
Comment (pi0) [inline packages/utils/src/ignore.js]: If this util is only used for builder (build time) i suggest moving it to the builder package. To reduce utils install size.

File: packages/builder/src/builder.js [hunk 5]
```diff
@@ -130,6 +134,23 @@ export default class Builder {
...
+  async resolveFiles(dir, cwd = this.options.srcDir) {
+    return this.ignore.filter(await glob(`${dir}/**/*.{${this.supportedExtensions.join(',')}}`, {
+      cwd,
+      ignore: this.options.ignore
+    }))
+  }
+
+  async resolveRelative(dir) {
+    const resources = []
+    for (const file of await this.resolveFiles(dir)) {
```
Comment (pi0) [inline packages/builder/src/builder.js]: Can we simplify it with  then + array.map?

File: packages/vue-app/template/store.js [hunk 6]
```diff
@@ -8,74 +8,28 @@ let storeData = {}
...
+  storeData = require('@/<%= dir.store %>/<%= s.src %>')
+  <% return true }}) %>
 
   // If store is not an exported method = modules store
   if (typeof storeData !== 'function') {
     // Store modules
     if (!storeData.modules) {
       storeData.modules = {}
     }
+    <% stores.forEach(s => {
```
Comment (pi0) [inline packages/vue-app/template/store.js]: Why not moving this part to builder/generateRoutes for making templates cleaner?

File: packages/vue-app/template/store.js [hunk 7]
```diff
@@ -8,74 +8,28 @@ let storeData = {}
 let files;
 
 void function updateModules() {
-  files = require.context('@/<%= dir.store %>', true, /^\.\/(?!<%= ignorePrefix %>)[^.]+\.(<%= extensions %>)$/)
```
Comment (pi0) [inline packages/vue-app/template/store.js]: Finally 🙈😂☺️❤️

File: packages/vue-app/template/store.js (hunk 6 above)
Comment (pi0) [reply packages/vue-app/template/store.js]: Builder:L394

File: packages/builder/package.json (hunk 1 above)
Comment (clarkdo) [reply packages/builder/package.json]: As @pi0 mentioned, maybe we should move ignore back to build for bundle size.

File: packages/vue-app/template/store.js (hunk 6 above)
Comment (clarkdo) [reply packages/vue-app/template/store.js]: It should, but I didn't touch the existing logic in template, I'll try.

File: packages/builder/src/builder.js [hunk 8]
```diff
@@ -377,15 +388,13 @@ export default class Builder {
...
-        cwd: this.options.srcDir,
-        ignore: this.options.ignore
-      })).forEach((f) => {
-        const key = f.replace(new RegExp(`\\.(${this.supportedExtensions.join('|')})$`), '')
-        if (/\.vue$/.test(f) || !files[key]) {
-          files[key] = f.replace(/(['"])/g, '\\$1')
+      const ext = new RegExp(`\\.(${this.supportedExtensions.join('|')})$`)
+      for (const page of await this.resolveFiles(this.options.dir.pages)) {
+        const key = page.replace(ext, '')
+        if (/\.vue$/.test(page) || !files[key]) {
```
Comment (ricardogobbosouza) [inline packages/builder/src/builder.js]: Change condition `||` to `&&`. See #4641

**Review by ricardogobbosouza — CHANGES_REQUESTED** [review]

File: packages/builder/src/builder.js (hunk 8 above)
Comment (ricardogobbosouza) [reply packages/builder/src/builder.js]: Sorry, the problem is not in `pages` but in` layouts`

Comment (ricardogobbosouza) [reply packages/builder/src/builder.js]: If try register layout in module doesn't work

```js
export default {
    modules: [
        {
            handler() {
                this.addLayout('./templates/my-layout.vue', 'default');
            }
        }
    ]
}
```

The problem is here:
```js
templateVars.layouts[name] = this.relativeToBuild(
    this.options.srcDir,
    file
);
```

Comment (clarkdo) [reply packages/builder/src/builder.js]: @ricardogobbosouza Gotcha, I'll try to figure out a proper fix.

Comment (clarkdo) [reply packages/builder/src/builder.js]: @ricardogobbosouza Fix can be reviewed in #4702 😄

File: packages/builder/src/builder.js [hunk 9]
```diff
@@ -318,14 +334,12 @@ export default class Builder {
...
       env: this.options.env,
       head: this.options.head,
-      middleware: fsExtra.existsSync(path.join(this.options.srcDir, this.options.dir.middleware)),
       store: this.options.store,
       globalName: this.options.globalName,
       globals: this.globals,
       css: this.options.css,
       plugins: this.plugins,
       appPath: './App.js',
-      ignorePrefix: this.options.ignorePrefix,
```
Comment (pi0) [inline packages/builder/src/builder.js]: Is this a breaking change?

File: packages/builder/src/builder.js [hunk 10]
```diff
@@ -392,16 +403,14 @@ export default class Builder {
     } else if (this._nuxtPages) {
       // Use nuxt.js createRoutes bases on pages/
       const files = {}
-      ;(await glob(`${this.options.dir.pages}/**/*.{${this.supportedExtensions.join(',')}}`, {
-        cwd: this.options.srcDir,
-        ignore: this.options.ignore
-      })).forEach((f) => {
-        const key = f.replace(new RegExp(`\\.(${this.supportedExtensions.join('|')})$`), '')
+      const ext = new RegExp(`\\.(${this.supportedExtensions.join('|')})$`)
```
Comment (pi0) [inline packages/builder/src/builder.js]: Nice cleanup 👍

**Review by pi0 — APPROVED** [review]

**Comment by manniL** [comment]
@clarkdo Doc PR?

**Comment by clarkdo** [comment]
@manniL Working on it

**Comment by manniL** [comment]
@clarkdo Should we deprecate ignore / ignorePrefix in 2.x? 🤔

**Comment by clarkdo** [comment]
@manniL As discussed before, it would be better to deprecate it in 3.0

**Pull request merged**